Add order metadata to intent before IPP capture

anu-rock opened the pull request:
Fixes #3228

### Background

A few columns in the list of transactions on **Payments → Transactions** page depend on order-related metadata attached to payment intents of corresponding transactions. Currently, Woo Mobile app supplies the metadata as seen here https://github.com/woocommerce/woocommerce-ios/issues/4077.

It is desirable to move the metadata responsibility from mobile app to plugin to:

1. Make plugin the single source of truth for metadata (mobile app cannot set `order_key`, so it'll need to be supplied by the plugin anyway).
2. Make the mobile app code DRY.

### Changes proposed in this Pull Request

Set metadata on intent during payment capture. The `update_intention` API may be used similar to how it's used in UPE's process_payment. (see line 371 below)

https://github.com/Automattic/woocommerce-payments/blob/ddbf9e4323bf85da650bbee9e49edef97fce8d98/includes/payment-methods/class-upe-payment-gateway.php#L365-L375

#### Testing Instructions

The thing that needs testing is a REST endpoint. There are two ways to test: Postman and Woo Mobile app. Both methods are dependent on 1331-gh-Automattic/woocommerce-payments-server.

##### Postman (recommended)

* Create a new order in WooCommerce with "Cash on delivery" as the payment method.

* Create a payment intent in Stripe Dashboard at `https://dashboard.stripe.com/test/connect/accounts/<account_id>/payments/new`. Make sure to have the "Capture funds later" option selected under "Manually enter card information."

* On the intent details page `https://dashboard.stripe.com/test/connect/accounts/<account_id>/payments/<intent_id>`, scroll down to the "Metadata" section. Add a couple of key-value pairs, eg. customer_name = John Doe, data_from_mobile = Dummy Text. We'll need this to verify backward compatibility later.

<img width="1395" alt="image" src="https://user-images.githubusercontent.com/1288616/143211149-8f50e3a7-8e91-401e-b263-efbfbf3ee908.png">

* Make a POST request to `/payments/orders/<order_id>/capture_terminal_payment`. Refer the [docs](https://github.com/woocommerce/woocommerce/wiki/Getting-started-with-the-REST-API) to set up authorization correctly.

<img width="1552" alt="image" src="https://user-images.githubusercontent.com/1288616/143206523-bf4d0452-b00f-4c67-92b6-ec15b87f13dc.png">

* Refresh the intent details page and verify whether new metadata key-value pairs have been added:
  * The original value of `customer_name` should be preserved, i.e. "John Doe." It should not be the one specified in the customer name field while creating the order in step 1. (backward compatibility check)
  * `data_from_mobile` should still be present. (backward compatibility check)
  * More pairs should have been added by the `capture_terminal_payment` API, such as order_id, order_key, etc.

##### Woo Mobile app (closest to the real world)

* Set up the app for testing as per the guide pdjVC1-1r-p2.

* Create a new order in WooCommerce with "Cash on delivery" as the payment method.

* Proceed to capture payment for the above order via the mobile app.

* Visit the intent details page `https://dashboard.stripe.com/test/connect/accounts/<account_id>/payments/<intent_id>` and verify that the metadata is present and as expected.
 
In the mobile app testing method, it's a bit hard to validate backward compatibility here as metadata supplied by the mobile app is usually identical to the one added by WooCommerce Payments. The Postman method is better for the backward compatibility checks.

### Approaches considered

> **Edit:** The following approaches do not take into consideration the merging of metadata extracted from intent and order. See this [comment](https://github.com/Automattic/woocommerce-payments/pull/3342#issuecomment-967021974) for revised pseudo code.

1. Call `update_intention` before capturing charge.
2. Call `update_intention` after capturing charge.

Both approaches need to be backward-compatible with the mobile app. That is, older versions of the app that set metadata will co-exist with newer versions of the app that set no metadata.

#### `update_intention` before capturing charge

Updating an intent before capturing it feels intuitive, like the right thing to do. Here's how pseudo code changes will look like in `WC_Payment_Gateway_WCPay::capture_charge`:

```php
$intent = $this->payments_api_client->get_intent( $id );

if ( empty ( $intent->get_metadata() ) ) {
    $updated_intent = $this->payments_api_client->update_intention( $id, $metadata );
}

$this->payments_api_client->capture_intention( ... );
```

The empty metadata check is required for backward compatibility, as noted earlier. In the worst case, we'll incur 3 Stripe API calls — get intent, update intent, and capture intent.

#### `update_intention` after capturing charge

This is an optimization over the previous approach in terms of the number of network calls. I've verified that Stripe API allows updating an intent after capturing.

```php
$intent = $this->payments_api_client->capture_intention( ... );

if ( empty ( $intent->get_metadata() ) ) {
    $updated_intent = $this->payments_api_client->update_intention( $id, $metadata );
}
```

---

Both approaches above will need the following changes:

* Add metadata field and get_metadata method in `WC_Payments_API_Intention`.
This should be straightforward.

* Refactor `update_intention` to support updating individual or a subset of properties.
Currently, our implementation expects 'amount' as a mandatory property. When we try to update metadata along with amount and some other properties, the Stripe API rejects the request as it doesn't allow updating amount on an intent with status 'requires_capture'. Catch-22. So, it is impossible to update metadata with or without specifying an amount. The Stripe API doesn't have such expectations.

-------------------

- [x] Added changelog entry
- [x] Covered with tests
- [x] Tested on mobile

**Post merge**

- [ ] Added testing instructions to the [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) (does not apply)

anu-rock committed:
Mark the desired place to update intent with metadata

kalessil commented:
IMO, `$intent = $this->api_client->get_intent( $intent_id );` is making more sense.
Ideally we should take merge the intent meta-data it already has with metadata generated in `\WC_Payments_API_Client::request_with_level3_data` - then we'll need only client changes and this will be compatible with any changes on the mobile app side.

What do you think?

anu-rock commented:
@kalessil Are you proposing something on the lines of this?

includes/class-wc-payment-gateway-wcpay.php
```php
$intent            = $this->payments_api_client->get_intent( $id );
$existing_metadata = $intent->get_metadata();
$new_metadata      = $this->get_metadata_from_order( $order, $payment_type );
$merged_metadata   = $this->merge_metadata( $existing_metadata, $new_metadata );

$this->payments_api_client->capture_intention( ..., $merged_metadata ); // internally calls `request_with_level3_data`
```

includes/wc-payment-api/class-wc-payments-api-client.php
```php
private function request_with_level3_data( $params, $api, $method, $is_site_specific = true ) {
    // Existing code
    ...
    // Send merged metadata along with capture intent request
}
```

If my understanding is indeed correct, this will effectively reduce 1 network call. But I see a couple of issues:

* The single responsibility principle will be violated for `request_with_level3_data`.
* The Stripe API doesn't allow updating an intent while capturing it.

Apologies if I misunderstood.

At any rate, I missed the merging aspect earlier. So, thanks for suggesting it. If we go with the **update_intention before capturing charge** approach, it will probably look like this:

```php
$intent            = $this->payments_api_client->get_intent( $id );
$existing_metadata = $intent->get_metadata();
$new_metadata      = $this->get_metadata_from_order( $order, $payment_type );
$merged_metadata   = $this->merge_metadata( $existing_metadata, $new_metadata );

$updated_intent = $this->payments_api_client->update_intention( $id, $merged_metadata );

$this->payments_api_client->capture_intention( ... );
```

But that will mean it's 3 network calls always.

anu-rock commented:
> Both approaches above will need the following changes:
> 
>     * Add metadata field and get_metadata method in `WC_Payments_API_Intention`.
>       This should be straightforward.
> 
>     * Refactor `update_intention` to support updating individual or a subset of properties.
>       Currently, our implementation expects 'amount' as a mandatory property. When we try to update metadata along with amount and some other properties, the Stripe API rejects the request as it doesn't allow updating amount on an intent with status 'requires_capture'. Catch-22. So, it is impossible to update metadata with or without specifying an amount. The Stripe API doesn't have such expectations.

@kalessil Also, what do you think about this?

kalessil commented:
@anu-rock:

> But that will mean it's 3 network calls always.

Something like this, yes. Adding one more API call will not hurt here, IMO - we have no alternatives anyway and backward compatibility for the mobile app is important.

> Add metadata field and get_metadata method in `WC_Payments_API_Intention`

Sounds good

> Refactor `update_intention` to support updating an individual or a subset of properties.

Can we have a specialized method `update_intention_metadata` instead, that'd better follow the single responsibility approach?

anu-rock committed:
Add metadata to intent during IPP charge capture

Update an IPP intent just before capturing it with metadata collated
from mobile app and WC order object.

In order not to impact existing functionality, we've created a new API
method to update intention.

anu-rock committed:
Add the missing payment_type variable

anu-rock committed:
Fix capture charge tests that broke due to metadata change

anu-rock committed:
Add a test to verify metadata in capture charge

anu-rock committed:
Merge branch 'develop' into fix/3228-ipp-add-metadata-to-intent-during-capture

anu-rock committed:
Attempt to fix syntax error due to trailing comma

dpaun1985 commented on tests/unit/test-class-wc-payment-gateway-wcpay.php:
```diff
@@ -812,6 +814,24 @@ public function test_capture_charge_success() {
 		$order->update_meta_data( '_intention_status', 'requires_capture' );
 		$order->update_status( 'on-hold' );
 
+		$mock_intent = new WC_Payments_API_Intention(
```
most of the test methods looks the same. I wonder if we can use a dataProivder. I know that the test methods were in place, but if we can use a data provider we can do a refactor.

anu-rock replied:
True. Thanks for pointing it out. I'll see if I could do the refactor.

dpaun1985 reviewed (commented):
Looks good overall. Great job! Left a few small remarks.

anu-rock commented on includes/wc-payment-api/class-wc-payments-api-client.php:
```diff
@@ -310,6 +310,67 @@ public function update_intention(
+		if ( ! empty( $level3 ) ) {
+			$request['level3'] = $level3;
+		}
+		if ( '' !== $selected_upe_payment_type ) {
```
Yes, that makes sense. Another way to go about it could be to modify the original `update_intention` to support our use case in a non-disruptive way. Something similar was also proposed here 1331-gh-Automattic/woocommerce-payments-server#discussion_r752249034. I'll look into it 👍

anu-rock replied:
This change is no longer required as `update_intention_unrestricted` has been replaced by `update_intention_metadata`.

anu-rock committed:
Add type declarations to the new methods

anu-rock commented on includes/wc-payment-api/class-wc-payments-api-client.php:
```diff
@@ -310,6 +310,67 @@ public function update_intention(
+	 * @return WC_Payments_API_Intention
+	 * @throws API_Exception - Exception thrown on intention creation failure.
+	 */
+	public function update_intention_unrestricted(
```
Done 👍

anu-rock commented on includes/wc-payment-api/models/class-wc-payments-api-intention.php:
```diff
@@ -247,4 +258,13 @@ public function get_last_payment_error() {
+	 *
+	 * @return array
+	 */
+	public function get_metadata() {
```
Done 👍

anu-rock committed:
Revert "Add type declarations to the new methods"

This reverts commit 1c9047e3f849436d6da94faa6ff0d3d336d4acb4.

anu-rock commented:
@dpaun1985 I reverted the parameter type declarations as that resulted in plenty of errors:  
https://github.com/Automattic/woocommerce-payments/actions/runs/1503746465

PHP linter won't let me use nullable types either (eg. `?int`) as it's not supported in PHP < 8.

dmvrtx commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -1685,6 +1685,23 @@ public function capture_charge( $order ) {
 		$currency                 = WC_Payments_Utils::get_order_intent_currency( $order );
 
 		try {
+			$intent_id    = $order->get_transaction_id();
+			$intent       = $this->payments_api_client->get_intent( $intent_id );
+			$payment_type = $this->is_payment_recurring( $order->get_id() ) ? Payment_Type::RECURRING() : Payment_Type::SINGLE();
+
+			$metadata_from_intent = $intent->get_metadata(); // mobile app may have set metadata.
+			$metadata_from_order  = $this->get_metadata_from_order( $order, $payment_type );
+			$merged_metadata      = array_merge( $metadata_from_order, $metadata_from_intent ); // prioritize metadata from mobile app.
+
+			$this->payments_api_client->update_intention_unrestricted(
+				$intent_id,
+				null,
+				null,
+				null,
+				null,
+				$merged_metadata
+			);
+
```
I wonder if it would be possible to extract this into a separate method, say in the `Order` class. This code flow requires `$order` as an input and updates related intent, so maybe having something like `$order->refresh_intention_metadata` would make it look cleaner in this method and allow for re-use whenever needed.

dmvrtx commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -1685,6 +1685,23 @@ public function capture_charge( $order ) {
+			);
+
 			$intent = $this->payments_api_client->capture_intention(
 				$order->get_transaction_id(),
```
In the current implementation you have it in the `$intent_id`

anu-rock replied:
Done ✅ Thanks for spotting it.

anu-rock committed:
Reuse the intent id extracted earlier

anu-rock committed:
Refactor update_intention_unrestricted as update_intention_metadata

This change moves us from a generalized method to a more specific one.
This has a couple of advantages:
* avoids code duplication and maintainability issues
* saves us from over-engineering

anu-rock committed:
Fix metadata merging bug

The bug: When there was no metadata in the given intent, array_merge returned null.

The issue is likely caused by a wrong type returned by
WC_Payments_API_Intention::get_metadata() when the upstream intent
has an empty array set for its metadata property.

anu-rock committed:
Remove the TODO in capture_charge's catch

Re-using intent might not be possible. This branch is recovering from
the exception which occurred in a separate code branch and intent
might not be available yet.

dmvrtx commented on includes/wc-payment-api/class-wc-payments-api-client.php:
```diff
@@ -310,6 +310,30 @@ public function update_intention(
+	 */
+	public function update_intention_metadata(
+		$intention_id,
+		$metadata = []
```
Is this default value required here? I believe it is better to throw a tantrum here and complain about missing `$metadata` rather than silently reset existing data to oblivion.

kalessil replied:
+1. Can we make `$metadata` a required parameter?

anu-rock replied:
Fixed in 6702736 ✅

kalessil commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -1685,8 +1685,21 @@ public function capture_charge( $order ) {
+
+			$metadata_from_intent = $intent->get_metadata(); // mobile app may have set metadata.
+			$metadata_from_order  = $this->get_metadata_from_order( $order, $payment_type );
+			$merged_metadata      = array_merge( (array) $metadata_from_order, (array) $metadata_from_intent ); // prioritize metadata from mobile app.
```
Nice!

kalessil reviewed (commented):
Yay, looks great! Left a couple of comments.

anu-rock committed:
Make metadata a required param

dmvrtx committed:
Merge branch 'develop' into fix/3228-ipp-add-metadata-to-intent-during-capture

dmvrtx reviewed (approved):
Looks good to me! Like the detailed testing instructions 🙂 

Let's fix a warning in tests and ship it!

anu-rock committed:
Rectify method name in mocks for intent update tests

anu-rock committed:
Merge branch 'fix/3228-ipp-add-metadata-to-intent-during-capture' of https://github.com/Automattic/woocommerce-payments into fix/3228-ipp-add-metadata-to-intent-during-capture

anu-rock commented on tests/unit/test-class-wc-payment-gateway-wcpay.php:
```diff
@@ -812,6 +814,24 @@ public function test_capture_charge_success() {
+		$this->mock_api_client->expects( $this->once() )->method( 'get_intent' )->with( $intent_id )->will(
+			$this->returnValue( $mock_intent )
+		);
+		$this->mock_api_client->expects( $this->once() )->method( 'update_intention_unrestricted' )->will(
```
Holy cow! How did I miss that 😕   
Thanks for spotting! You have an eagle's eyes.

anu-rock replied:
Done ✅

anu-rock committed:
Update changelog

kalessil commented on includes/wc-payment-api/class-wc-payments-api-client.php:
```diff
@@ -313,6 +313,30 @@ public function update_intention(
+	 * @return WC_Payments_API_Intention
+	 * @throws API_Exception - Exception thrown on intention creation failure.
+	 */
+	public function update_intention_metadata(
```
Can we make the method signature a one-liner?

anu-rock replied:
Done ✅

kalessil commented on includes/wc-payment-api/class-wc-payments-api-client.php:
```diff
@@ -313,6 +313,30 @@ public function update_intention(
+		$intention_id,
+		$metadata
+	) {
+		$request = [
```
I'd prefer to dispatch `[ 'metadata' => $metadata ]` right into the call where it's been used.

anu-rock replied:
I think giving this array a name makes it instantly understandable by new folks. I prefer to use variables to add comments to my code, wherever suitable. Let me know if you'd still have it the other way.

kalessil replied:
Ok, go ahead then.

kalessil commented on tests/unit/test-class-wc-payment-gateway-wcpay.php:
```diff
@@ -1191,26 +1360,28 @@ public function test_capture_charge_expired() {
-		$this->assertEquals( $order->get_status(), 'cancelled' );
+		$this->assertContains( 'successfully captured', $latest_wcpay_note->content );
+		$this->assertContains( wc_price( $order->get_total() ), $latest_wcpay_note->content );
+		$this->assertEquals( $order->get_meta( '_intention_status', true ), 'succeeded' );
```
`assertSame` =)

anu-rock replied:
Gosh, how did I miss this despite constant prodding from you 😄  
Done ✅

kalessil reviewed (approved):
Did review only (not tests) and LGTM! Also left minor improvements suggessions.

anu-rock committed:
Finishing touches

* Beautify method declaration
* s/AssertEquals/AssertSame

anu-rock committed:
Merge branch 'develop' into fix/3228-ipp-add-metadata-to-intent-during-capture

anu-rock committed:
Merge branch 'develop' into fix/3228-ipp-add-metadata-to-intent-during-capture
